enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,210 @@
+
+That being said, to address this issue we propose to document a simple workaround:
+
+1. Enable the `CSIMigrationNode` *FeatureSet*.
+1. Downgrade.
+
+As stated above, the CSI migration *FeatureSets* are not operational once CSI migration becomes GA. However, they will be carried over during the downgrade and they will be correctly applied once the system is downgraded.
+
+### Post-GA
+
+CSI migration *FeatureSets* can be removed from OCP API **one** release after CSI migration becomes GA.
+
+### Risks and Mitigations
+
+Although this three-phased approach does what we need, it has some drawbacks:
+
+1. Having operators ignoring certain *FeatureSets* is not usual and is error-prone. Fortunately we only need to introduce the skipping in MCO.
+
+## Design Details
+
+This is what needs to be done across different support phases:
+
+1. Tech Preview:
+
+* openshift/api
+  * Introduce two new *FeatureSets*: `CSIMigrationNode` and `CSIMigrationControlPlane`
+* Machine Config Operator
+  * Bump openshift/api in order to get the *FeatureSets* above
+  * Introduce a patch to ignore the `CSIMigrationControlPlane` *FeatureSet*
+* Kubernetes Scheduler Operator
+  * Bump openshift/api in order to get the *FeatureSets* above
+* Kubernetes Controller Manager Operator
+  * Bump openshift/api in order to get the *FeatureSets* above
+
+Other operators, like the Kubernetes API Server Operator, may have their openshift/api library bumped. However, that is not strictly necessary as their operands don't need to enable the CSI migration feature flags.
+
+2. GA
+
+Nothing needs to be done.
+
+3. GA + 1 release
+
+* openshift/api
+  * Remove CSI migration *FeatureSets*: `CSIMigrationNode` and `CSIMigrationControlPlane`
+* Machine Config Operator
+  * Bump openshift/api
+  * Remove the skip added for Tech Preview
+* Other operators
+  * Bump openshift/api
+
+### Test Plan
+
+#### E2E jobs
+
+We want E2E jobs for all migrated plugins ready at **Tech Preview** time.
+
+For each E2E job:
+
+1. Install an OCP cluster.
+1. Enable the feature gate. Don't worry about the order in which components will apply the feature flags because at this point there are no volumes created.
+1. Wait until MCO and control-plane operators restart.
+1. Run E2E tests for in-tree volume plugins.
+1. Disable the feature gate. Again, don't worry about the ordering.
+1. Run E2E tests again.
+
+In addition to that, as a strech goal, we want a separate job that:
+
+1. Runs a StatefulSet.
+1. Enables the migration *FeatureSets* in the right order.
+1. Wait for all components to have the right feature flags.
+1. Checks if the StatefulSet survives.
+
+Once CSI migration is GA, we expect the regular upgrade jobs will cover upgrades from an OCP version with migration disabled to a version with migration enabled.
```

Comment:
Please add a note about downgrade. We expect that just installed cluster does not have any PVs/PVC. Therefore, even though downgrade job downgrades the components in a wrong order, it will not have any bad effect.

Reply:
Well, see the note above. Any downgrade job in CI will likely fail. Luckily, we don't have downgrade jobs (yet).